lib/device-connections-factory.js

```diff
@@ -0,0 +1,169 @@
+import _ from 'lodash';
+import net from 'net';
+import B from 'bluebird';
+import { logger } from 'appium-support';
+import { utilities } from 'appium-ios-device';
+import { checkPortStatus } from 'portscanner';
+
+const log = logger.getLogger('iProxy');
+
+class iProxy {
+  constructor (udid, localport, deviceport) {
+    this.localport = parseInt(localport, 10);
+    this.deviceport = parseInt(deviceport, 10);
+    this.udid = udid;
+    this.serverSocket = null;
+  }
+
+  async start () {
+    if (this.serverSocket) {
+      return;
+    }
+    this.serverSocket = net.createServer(async (connection) => {
+      try {
+        const socket = await utilities.connectPort(this.udid, this.deviceport);
+        socket.on('close', connection.destroy);
+        socket.on('error', log.error);
+        connection.on('close', socket.destroy);
+        connection.on('error', log.error);
+        connection.pipe(socket);
+        socket.pipe(connection);
+      } catch (e) {
+        log.warn(e.message);
+        connection.destroy();
+      }
+    });
+    const status = new B((resolve, reject) => {
+      this.serverSocket.once('listening', resolve);
+      this.serverSocket.once('error', reject);
+    });
+    this.serverSocket.listen(this.localport);
+    await status;
+  }
+
+  quit () {
+    if (!this.serverSocket) {
+      return;
+    }
+    this.serverSocket.close();
+    this.serverSocket = null;
+  }
+}
+
+class DeviceConnectionsFactory {
+  constructor () {
+    this._connectionsMapping = {};
+  }
+
+  _toKey (udid = null, port = null) {
+    return `${udid || ''}:${port || ''}`;
+  }
+
+  listConnections (udid = null, port = null, strict = false) {
+    if (!udid && !port) {
+      return [];
+    }
+
+    const expectedKey = this._toKey(udid, port);
+    // `this._connectionMapping` keys have format `udid:port`
+    // the `strict` argument enforces to match keys having both `udid` and `port`
+    // while in non-strict mode keys having any of these are going to be matched
+    return _.keys(this._connectionsMapping)
+      .filter((key) => strict
+        ? key.startsWith(expectedKey) && key.endsWith(expectedKey)
+        : key.startsWith(expectedKey) || key.endsWith(expectedKey));
+  }
+
+  async requestConnection (udid, port, options = {}) {
+    if (!udid || !port) {
+      log.warn('Either device UDID or local port is unset. ' +
```

Comment:
Here was raised. let me check

Reply:
Shown udid and port. hm... udid is undefined => `Either device UDID 'undefined' or local port '8100' is unset.`

Reply:
changed the logging